server/base/src/main/java/org/apache/accumulo/server/metrics/PerTableMetrics.java

```diff
@@ -0,0 +1,297 @@
+    }
+
+  }
+
+  public PerTableMetrics(ServerContext context, ActiveTableIdTracker activeTableIdTracker) {
+    this.context = context;
+    this.activeTableIdTracker = activeTableIdTracker;
+    this.activeTableIdTracker.addChangeListener(this);
+  }
+
+  /**
+   * This method exist so this class can log using the logger of the subclass.
+   */
+  protected abstract Logger getLog();
+
+  /**
+   * Subclasses should implement this method to create a TableMetrics object that will be used in
+   * the case when per table metrics are disabled. The object returned by this method will alway be
+   * returned by {@link #getTableMetrics(TableId)} no matter what the table id is.
+   *
+   * @param registry register an meters for the table metrics in this registry
+   * @param meters a consumer that accepts meters to be removed from the registry when the table
+   *        metrics object is discarded. Currently this consumer does nothing with the meters, its
+   *        passed for consistency with
+   *        {@link #newPerTableMetrics(MeterRegistry, TableId, Consumer, List)}
+   * @param tags currently an empty collection of tags, this is passed for consistency with
+   *        {@link #newPerTableMetrics(MeterRegistry, TableId, Consumer, List)}
+   * @return a new object that will be cached and later returned by
+   *         {@link #getTableMetrics(TableId)}
+   */
+  protected abstract T newAllTablesMetrics(MeterRegistry registry, Consumer<Meter> meters,
+      List<Tag> tags);
+
+  /**
+   *
+   * Subclasses should implement this method to create per table table metrics objects. This method
+   * is called in the case where per table metrics are enabled. These objects will be cached and
+   * returned by {@link #getTableMetrics(TableId)}. Table metrics object in the cache that are no
+   * longer needed will be automatically removed when the table is deleted or this server has not
+   * hosted the table for a bit.
+   *
+   * @param registry register an meters for the table metrics in this registry
+   * @param meters a consumer that accepts meters to be removed from the registry when the per table
+   *        metrics object is discarded.
+   * @param tags returns a list with a single tag in it which is the tableId. These tags should be
+   *        used when registering meters
+   * @return a new object that will be cached and later returned by
+   *         {@link #getTableMetrics(TableId)}
+   */
+  protected abstract T newPerTableMetrics(MeterRegistry registry, TableId tableId,
+      Consumer<Meter> meters, List<Tag> tags);
+
+  private TableMetricsInfo<T> getOrCreateTableMetrics(TableId tableId) {
+    Preconditions.checkState(activeTableIdTracker.isPerTableMetricsEnabled());
+    return perTableMetrics.computeIfAbsent(tableId, tid -> {
+      List<Meter> meters = new ArrayList<>();
+      T tableMetrics = newPerTableMetrics(registry, tableId, meters::add,
+          List.of(Tag.of(TABLE_ID_TAG_NAME, tid.canonical())));
```

Comment:
I'm thinking it might be more useful that have a tableName tag and use the table name instead of the tableId.